### ISSUE TITLE
Rerouting doesn't work when Route was calculated with Arabic or Hebrew locale

### ISSUE THREAD

**Issue by SebastianOsinski** [mapbox/mapbox-navigation-ios#2336]
**Mapbox Navigation SDK version: 0.38.1**

Today we noticed problem with rerouting - when user wanders off current route, rerouting is not  triggered. We experience it only when running app with Arabic locale. When our app is run with English locale, then everything works as expected. 

Additionally, when we run app in Arabic, but overwrite locale to English (or any other LTR language we tested) before making request for route, rerouting still works as expected when we start navigation with this route:

```swift
let options = NavigationRouteOptions(waypoints: waypoints, profileIdentifier: profileIdentifier)
options.locale = Locale("en") // Locale("de"), Locale("pl") also works fine
directions.calculate(options) { _, routes, error in
    if let route = routes.first {
        // pass route to NavigationService and start navigation
    }
}
```

We investigated it further by following code execution and the problem seems to be in `userIsOnRoute` method in `RouteController` which always returns false:
 https://github.com/mapbox/mapbox-navigation-ios/blob/cae1c93980937b58a2f138517d9ad00540bed69a/MapboxCoreNavigation/RouteController.swift#L382

`status.routeState` always returns invalid when Route was calculated with `ar` or `he`  locale.
`status` is returned by `getStatusForTimestamp` method from `MBNavigator`. `MBNavigator` comes from `MapboxNavigationNative` which is closed-source so we can't debug this issue any further.

It looks like `MBNavigator` either might have some problem with working with Routes for RTL languages or the response from Directions API for RTL languages lacks some information

**Comment by SebastianOsinski** [comment]
We just migrated to MapboxNavigation 1.0.0-alpha.1 (along with MapboxDirections 1.0.0-alpha.1) to see what would happen and rerouting for Arabic works fine now. Looks like the problem was caused by something in Mapbox Directions 0.30.0

**Comment by 1ec5** [comment]
In v0.38, we were passing Directions API responses directly to MapboxNavigationNative’s MBNavigator class, whereas in v1.0 we’re roundtripping the response through Swift model objects. It’s surprising that the former approach would be less reliable than the latter in these locales. I wonder if there’s an encoding issue of some sort.

When `status.routeState` was `invalid`, was `status.stateMessage` set to anything other than an empty string?

**Comment by SebastianOsinski** [comment]
@1ec5 Thanks for quick response. I checked `status.stateMessage` on `v0.38` and its value is `"route voiceLocale is not a string."`.

**Comment by 1ec5** [comment]
Ah, that’s a good find. The Directions API sets `voiceLocale` to `null` when the language passed in with the `language` parameter is supported by the Directions API (OSRM Text Instructions behind the scenes) but not by the Voice API (Amazon Polly behind the scenes). That would be the case with any language marked “Uses VoiceOver” in [this table](https://docs.mapbox.com/ios/navigation/overview/localization-and-internationalization/#supported-languages), although [the Directions API documentation](https://docs.mapbox.com/api/navigation/#route-object) doesn’t spell out this edge case.

MBNavigator apparently does some validation on the route JSON that RouteController gives it. It expects `voiceLocale` to be a string but doesn’t allow for it to be `null`. This issue affected navigation SDK v0.38 but not v1.0 because MapboxDirections v1.0 round-trips the JSON to a Route using JSONDecoder and back to a JSON using JSONEncoder. JSONEncoder never reencodes the `null` value because [`KeyedEncodingContainer.encodeIfPresent(_:forKey:)`](https://github.com/mapbox/mapbox-directions-swift/blob/fe77075932d469473237bb5346dfcc5909124886/Sources/MapboxDirections/DirectionsResult.swift#L84) is a no-op if the value is `nil`.

I’ve reported this issue internally to the team responsible for MapboxNavigationNative. The issue doesn’t affect v1.0.0-alpha.1, so technically we’ve already worked around it, but I’m going to leave this issue open until we upgrade to a version of MapboxNavigationNative that can cope with a `null` value, just in case.

**Comment by 1ec5** [comment]
> This issue affected navigation SDK v0.38 but not v1.0 because MapboxDirections v1.0 round-trips the JSON to a Route using JSONDecoder and back to JSON using JSONEncoder. JSONEncoder never reencodes the `null` value because [`KeyedEncodingContainer.encodeIfPresent(_:forKey:)`](https://github.com/mapbox/mapbox-directions-swift/blob/fe77075932d469473237bb5346dfcc5909124886/Sources/MapboxDirections/DirectionsResult.swift#L84) is a no-op if the value is `nil`.

The Directions API intentionally sets this property to `null` to signal to the navigation SDK that it shouldn’t rely on the Voice API to utter the instructions. The navigation SDK should round-trip the `null` to keep the instructions from being read in the wrong language. Indeed, when I run a navigation SDK–powered application in Vietnamese, the instructions are read by an Amazon Polly voice, whereas I’d expect something client-side that could potentially be a third-party text-to-speech engine. We would need to make the fix once the MapboxNavigationNative bug is resolved.

**Comment by SebastianOsinski** [comment]
@1ec5 Thank you for this thorough explanation! Is there some estimate or public bug tracker for MapboxNagiationNative where we can check progress of this fix?

**Comment by 1ec5** [comment]
Unfortunately there isn’t a public bug tracker for MapboxNavigationNative. The fix is about to land in a new version of MapboxNavigationNative, but the navigation SDK currently depends on a very old version. #2319 would upgrade to a newer version; I’ll update that PR as soon as the fix is out in MapboxNavigationNative, but we’ll still need more testing before we can land it.

If you need a workaround in v0.38._x_, you could delete the `voiceLocale` key from the [public but undocumented `Route.json` dictionary](https://github.com/mapbox/mapbox-directions-swift/blob/65857d14080afa1e5463e0dbadd9dec23e66a6f3/Sources/MapboxDirections/MBDirectionsResult.swift#L14) just before starting turn-by-turn navigation in NavigationViewController. You’d need to do the same thing every time the navigation SDK reroutes the user, perhaps in [`NavigationViewControllerDelegate.navigationViewController(_:didRerouteAlong:)`](https://docs.mapbox.com/ios/api/navigation/0.38.3/Protocols/NavigationViewControllerDelegate.html#/c:@M@MapboxNavigation@objc%28pl%29MBNavigationViewControllerDelegate%28im%29navigationViewController:didRerouteAlongRoute:) or [`NavigationServiceDelegate.navigationService(_:didRerouteAlong:at:proactive:)`](https://docs.mapbox.com/ios/api/navigation/0.38.3/Protocols/NavigationServiceDelegate.html#/c:@M@MapboxNavigation@objc%28pl%29NavigationServiceDelegate%28im%29navigationService:didRerouteAlongRoute:at:proactive:).

> The navigation SDK should round-trip the `null` to keep the instructions from being read in the wrong language.

Tracking the instructions being read in the wrong language by the Voice API in mapbox/mapbox-directions-swift#415. A workaround in v1.0.0-alpha.1 or master would be to replace the default MapboxVoiceController with a RouteVoiceController when setting up the NavigationOptions.

**Comment by 1ec5** [comment]
#2319 upgraded to MapboxNavigationNative v9.0.4 to fix the root cause, while mapbox/mapbox-directions-swift#424 fixes the latent issue causing the instructions to be read in the wrong language.

**Comment by 1ec5** [comment]
Fixed in #2319 and mapbox/mapbox-directions-swift#424.
